src/core/shape_inference/include/utils.hpp

```diff
@@ -184,7 +184,14 @@ inline bool get_data_as_shape(
     T& shape,
     const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
     if (constant_data.count(idx)) {
-        shape = T(ov::opset1::Constant(constant_data.at(idx)).cast_vector<size_t>());
```

Comment:
This change is not exactly fix it replace one limitation with other.
Previous implementation works correctly for unsigned element types but wrong for signed and vice versa for new version (it can give false exception for unsigned values).

This change affect on all usages so there is some potential risk (if there some cases not covered by tests) that it will fail and previous implementation we can assume is proven by use (it is in project for some longer period of time).
So I would recommend make some custom change for one hot operator as this is really required only by this op.
Then in separate PR (I can think about during TopK review) to introduce new function which give correct functionality and use it reviewed operators. 
Then for old function maybe it can be just a wrapper for some new more generic function and with time usage can be review and if there any issue be replaced by new function.

Form other hand the new implementation pass tests so is big chance is good enough. Refactoring and introduction of new function can be done after this PR.

After merge of #14721, in TopK review I can do some clean-up for `get_data_as_shape` as I will need it TopK and could reuse it in tile (maybe slice also) and it looks could be beneficial for one hot.

Reply:
Changed to custom get_data_as_shape for OneHot.